---
title: "Single-Cell Analysis Workflow of cellxgene h5ad File"
output:
  html_document: default
params:
  knitr_eval: FALSE
  fileId: "e9e0243b-ed2b-4b6d-b049-7c7be43dd5b9"
  sample: "Disease.group"
  core: 10
  mem_gb: 60
vignette: >
  %\VignetteIndexEntry{Single-Cell Analysis Workflow of cellxgene h5ad File}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8} 
---

```{r setup}
knitr::opts_chunk$set(echo = TRUE, eval = params$knitr_eval)
```

This vignette follows single-cell analysis Workflow (workflow chapters of the “Orchestrating Single-Cell Analysis with Bioconductor” book) to analyze h5ad file from cellxgene dataset and tracks computing resources usage.

## Set up parameters

File id of h5ad file to analyze could be obtained using package `cellxgenedp`, and one example shows below.

```{r, echo = TRUE, message=FALSE}
library(cellxgenedp)
library(dplyr)
datasets(db()) |>
  filter(
    name == "Immunophenotyping of COVID-19 and influenza highlights the role of type I interferons in development of severe COVID-19"
  )

files(db()) |>
  filter(
    dataset_id == "0168ce6e-01a1-49e9-992e-b160c81b35a1",
    filetype == "H5AD"
  )
```

```{r, echo = TRUE, message=FALSE, eval=TRUE}
fileId <- params$fileId
```

Column to analyze needs to be specified from colData of the dataset.

```{r, echo = TRUE, message=FALSE, eval=TRUE}
sample <- params$sample
```

```{r, echo = TRUE, message=FALSE, eval=TRUE}
core <- params$core
```

```{r, echo = TRUE, message=FALSE, eval=TRUE}
## file id
fileId

## column selected for analysis
sample

## number of CPU cores detected
parallel::detectCores()

## number of CPU cores asked
core

## memory 
sprintf("%d GB", params$mem_gb) 
```

With parameters set up, workflow could be started. 

## Start for `Rcollectl` and download h5ad file

```{r cl_start, echo = TRUE, message=FALSE}
library(Rcollectl)
library(cellxgenedp)
library(dplyr)

## download h5ad file
h5ad_file <- files(db()) |>
  filter(file_id == "e9e0243b-ed2b-4b6d-b049-7c7be43dd5b9") |> files_download(dry.run = FALSE)

## start Rcollectl
id <- cl_start(basename(h5ad_file))
```

## Step data preparation 

```{r data_preparation, echo = TRUE, message=FALSE}
# record start time of each step
cl_timestamp(id, "data preparation")

## import h5ad file
library(zellkonverter)
sce <- readH5AD(h5ad_file, use_hdf5 = TRUE)

## pre-process data
library(SingleCellExperiment)
sce$sample <- colData(sce)$Name

library(EnsDb.Hsapiens.v86)
rowData(sce)$Chr <- mapIds(EnsDb.Hsapiens.v86, keys=rownames(sce), column="SEQNAME", keytype="GENEID")

names(assays(sce)) <- "counts"

saveRDS(sce, file = "sce_h5ad.rds")
```

## Step quality control

```{r quality_control, message=FALSE}
cl_timestamp(id, "quality control")

library(BiocParallel)
library(scater)

bpp <- MulticoreParam(as.integer(core))

sce <- unfiltered <- addPerCellQC(sce, BPPARAM=bpp, subsets=list(Mito=which(rowData(sce)$Chr=="MT")))
qc <- quickPerCellQC(colData(sce), batch=sce$sample, sub.fields="subsets_Mito_percent")

sce <- sce[,!qc$discard]
unfiltered$discard <- qc$discard

## Distribution of QC metrics in the dataset. Each point represents a cell and is colored according to whether it was discarded
gridExtra::grid.arrange(
    plotColData(unfiltered, x="sample", y="sum", colour_by="discard") +
        scale_y_log10() + ggtitle("Total count"),
    plotColData(unfiltered, x="sample", y="detected", colour_by="discard") +
        scale_y_log10() + ggtitle("Detected features"),
    plotColData(unfiltered, x="sample", y="subsets_Mito_percent",
                colour_by="discard") + ggtitle("Mito percent"),
    ncol=2
)

## Percentage of mitochondrial reads in each cell in the dataset compared to its total count. Each point represents a cell and is colored according to whether that cell was discarded
plotColData(unfiltered, x="sum", y="subsets_Mito_percent", colour_by="discard") + scale_x_log10()

## Normalization
sce <- logNormCounts(sce, size_factors = sce$sum)
summary(sizeFactors(sce))
```

## Step variance modeling

```{r variance_modeling, message=FALSE}
cl_timestamp(id, "variance modeling")

library(scran)
set.seed(101001)

dec <- modelGeneVarByPoisson(sce, block=sce$sample, BPPARAM=bpp)
top <- getTopHVGs(dec, n=5000)

## Per-gene variance as a function of the mean for the log-expression values in the dataset. Each point represents a gene (black) with the mean-variance trend (blue) fitted to the variances
par(mfrow=c(4,2))
par(mar=c(1,1,1,1))
blocked.stats <- dec$per.block
for (i in colnames(blocked.stats)) {
    current <- blocked.stats[[i]]
    plot(current$mean, current$total, main=i, pch=16, cex=0.5,
         xlab="Mean of log-expression", ylab="Variance of log-expression")
    curfit <- metadata(current)
    curve(curfit$trend(x), col='dodgerblue', add=TRUE, lwd=2)
}
```

## Step data integration

```{r data_integration, message=FALSE}
# This step takes longer time
cl_timestamp(id, "data integration")

library(batchelor)
library(BiocNeighbors)

set.seed(1010001)
merged <- fastMNN(sce, batch = sce$sample, subset.row = top,
                  BSPARAM=BiocSingular::RandomParam(deferred = TRUE), 
                  BNPARAM=AnnoyParam(), BPPARAM=bpp)

reducedDim(sce, 'MNN') <- reducedDim(merged, 'corrected')
metadata(merged)$merge.info$lost.var
```

## Step dimensionality reduction

```{r dimensionality_reduction, message=FALSE}
cl_timestamp(id, "dimensionality reduction")

library(uwot)
set.seed(1010001)
sce <- runUMAP(sce, dimred="MNN",
                    external_neighbors=TRUE, 
                    BNPARAM=AnnoyParam(),
                    BPPARAM=bpp,
                    n_threads=bpnworkers(bpp))
```

## Step clustering

```{r clustering, message=FALSE}
cl_timestamp(id, "clustering")

library(bluster)

set.seed(1010001)
colLabels(sce) <- clusterRows(reducedDim(sce, "MNN"), TwoStepParam(KmeansParam(centers=1000), NNGraphParam(k=5)))

table(colLabels(sce))

tab <- table(Cluster=colLabels(sce), sample=sce$sample)

## Heatmap of log10-number of cells in each cluster (row) from each sample (column)
library(pheatmap)
pheatmap(log10(tab+10), color=viridis::viridis(100))

scrambled <- sample(ncol(sce))

## UMAP plots of the dataset after merging. Each point represents a cell and is colored according to the assigned cluster (top) or the sample of origin (bottom)
gridExtra::grid.arrange(
    plotUMAP(sce, colour_by="label", text_by="label"),
    plotUMAP(sce[,scrambled], colour_by="sample")
)
```

## Step differential expression

```{r differential_expression, message=FALSE}
cl_timestamp(id, "differential expression")

markers <- findMarkers(sce, block = sce$sample, direction = 'up', lfc = 1, BPPARAM=bpp)

top.markers <- markers[["4"]]
best <- top.markers[top.markers$Top <= 10,]
lfcs <- getMarkerEffects(best)

## Heatmap of log2-fold changes for the top marker genes (rows) of cluster 4 compared to all other clusters (columns)
pheatmap(lfcs, breaks=seq(-5, 5, length.out=101))
```

## Step cell type classification

```{r cell_type_classification, message=FALSE}
cl_timestamp(id, "cell type classification")

se.aggregated <- sumCountsAcrossCells(sce, id=colLabels(sce), BPPARAM=bpp)

library(celldex)
hpc <- HumanPrimaryCellAtlasData()

library(SingleR)
anno.single <- SingleR(se.aggregated, ref = hpc, labels = hpc$label.main, assay.type.test="sum")
saveRDS(anno.single, file = "cell_type_classification.rds")
anno.single
```

## Stop `Rcollectl`

```{r cl_stop}
cl_stop(id)
```

## Plot collectl result

```{r plot_all}
path <- cl_result_path(id)
plot_usage(cl_parse(path))
```

## Plot collectl result with time stamp

```{r plot_timestamp}
path <- cl_result_path(id)
plot_usage(cl_parse(path)) +
  cl_timestamp_layer(path) +
  cl_timestamp_label(path) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
```

## `R` session information

```{r session_info, echo=FALSE}
## Session info
library("sessioninfo")
options(width = 120)
session_info()
```
